### DISCUSSION TITLE
Allow setting `indent` globally

### DISCUSSION THREAD

**Discussion by scgtrp** [helix-editor/helix#10363] [Ideas]
I currently have an incredibly long `~/.config/helix/languages.toml`, to which I append a copy of this for every language I use or encounter:

```
[[language]]
name = "..."
indent = { tab-width = 4, unit = "\t" }
auto-pairs = false
```

I discovered while writing this that I don't actually have to set `auto-pairs = false` in each language anymore (I don't think the global one existed/worked when I started using Helix?), so that's neat, thanks!

I would like to be able to do the same for `indent`. The desired behavior is that I can say "always indent with tabs, for any file type", and spaces are only used if I open an existing file that is already indented with spaces.

Potential issue with this: Some languages (e.g. makefiles) require a particular type of indentation as part of their syntax. These languages should probably have some sort of "no, really, use this even if the user said something else" flag in their `indent` configuration, to avoid emitting broken makefiles when the user has set space-indentation globally. Alternatively, the language-level `indent` could be split into `indent` and `default-indent`, where the order of precedence is `language.indent > global.indent > language.default-indent`.

**Comment by sploders101** [comment]
I would like to second this. I'm currently doing the exact same thing, and am still worried I run the risk of starting to use a new language that automatically starts using spaces for indentation without me noticing.

Not trying to get into the tabs vs spaces debate (just adding my objective reasoning for this need), but my personal opinion is that someone's personal indentation width preference should not be forced on everyone else. I have a lot of trouble reading 2-space indentation, so I like to render at 4, and tabs let me do that, so as a courtesy to others, I use tabs so they can view my code however they like. However, in today's development world, devs (or at least I) use so many languages on a day-to-day basis it's hard to keep track, and I'd love a global setting with language-specific override for languages with intolerant syntax (for example, Home Assistant's YAML parser really seems to hate tabs, so I should be able to override for YAML).

I may be willing to help with this in my spare time if someone could point me to the relevant parts of the project.

**Comment by pbrkr** [comment]
I think this is a key feature for users working on projects which have a strict coding style and use multiple programming languages.

Also note that simply enumerating indent overrides for all languages which helix currently knows about is not a one-off solution. This list would also have to be updated if new languages are added to Helix's default `languages.toml` file.

**Comment by dead10ck** [comment]
I'm actually personally against having a global setting for tab width. Each language has its own style, and I don't agree with the mindset of just ignoring idiomatic conventions. Just as one shouldn't write snake case in Java, or camel case in Rust, one shouldn't use an unconventional tab width.

Also, it's likely that if you really insisted on doing this, you're probably not talking about more than half a dozen languages, max. It's not like you'd notice some Prolog dialect was added and have to go add another entry.

**Comment by scgtrp** [comment]
I don't actually use Helix anymore (and will probably unsubscribe from this ticket), but as of when I stopped, I had 11: C++, C, Rust, JavaScript, TypeScript, HTML, YAML, Python, CSS, Markdown, and Bash. For the sort of personal use I was using it for, consistency with my own opinions was far more important than consistency with the opinions of some style guide.

The nice thing about optional settings is that you don't have to set them if they don't suit you :)

**Comment by gurgelgubbe** [comment]
I understand that I'm not the target audience as I'm pretty much a normie who enjoys tinkering with my computer, a non-dev, mostly janky scripts, configs and markdown documents. But I would also like some way of telling Helix: *unless told otherwise a tab consists of spaces, two spaces to be exact*.

**Comment by lucyColeclough** [comment]
> I'm actually personally against having a global setting for tab width. Each language has its own style, and I don't agree with the mindset of just ignoring idiomatic conventions. Just as one shouldn't write snake case in Java, or camel case in Rust, one shouldn't use an unconventional tab width.
> 
> Also, it's likely that if you really insisted on doing this, you're probably not talking about more than half a dozen languages, max. It's not like you'd notice some Prolog dialect was added and have to go add another entry.

My issue is that I don't want to write a config for every language.
If you're one who enjoys spaces then it wouldnt be as obvious of a problem.

**Comment by lucyColeclough** [comment]
in regards to #6652
I propose the notion that a global indent option would override any default configuration in helix
Then any option specified in the user~s `languages.toml` would then override this global option
This prevents the need to specify something like `indent.required= true` for each language, as it is assumed that if a user specifies a language in `languages.toml`, that it should override the global default
Perhaps behaviour for whether non specified fields should inherit from the global setting or the helix default
This could be a per language setting, either inherit from global or helix defaults, that way it would work with languages with no custom fields in  `languages.toml`

If going with that, then to increase reuse whilst resolving the issue posed in the openning comment of this thread, perhaps languages which are dependent on a specific indentation can have their inheritance set to the helix default, by default as opposed to the user specified global setting.

I do not have the time to change this my self and it is not an issue that would be desirable for me to see develop into a halt in existing work.

**Comment by Lilith-In-Starlight** [comment]
i think global options should override the default languages.toml *and* the global languages.toml, meaning they are only overriddeo by the languages.toml present in a project directory

This means that you can have

- Your customized global default
- Your customized edge cases for certain languages
- Your customized edge cases for certain projects

Which I think is an ideal amount of control

**Comment by Rudxain** [comment]
@dead10ck
> Each language has its own style, and I don't agree with the mindset of just ignoring idiomatic conventions.

That's why formatters and linters are a thing 🙃. And that's why the global setting should be just a default, overriden by whatever `indent.unit` is detected in a file or `languages.toml`.

> Just as one shouldn't write snake case in Java, or camel case in Rust, one shouldn't use an unconventional tab width.

`rustc` & `clippy` warn when using non-idiomatic naming-conventions. I would expect other linters (maybe even "intrusive" formatters) would go as far as to rename all symbols on "auto-fix"

**Reply by getawaywithrmdir** [reply]
I'm dying for a global indentation option.

I was even trying to figure out how to automatically run the ":indent-style" command every time I open a buffer for a clunky band-aid.

**Reply by kirawi** [reply]
I don't think we're planning to do this since we want the plugin system to handle this kind of stuff. But if you're not opposed to compiling from source, you would just replace the `self.indent_style` block in https://github.com/helix-editor/helix/blob/41dec92b0feb77719b07334b7a7f53152866e4e3/helix-view/src/document.rs#L1015-L1024

with

```rs
self.indent_style = IndentStyle::Tabs;
// Or
self.indent_style = IndentStyle::Spaces(n)
```

But don't remove the `if let Some ...` stuff

**Reply by Rudxain** [reply]
> compiling from source

About that, if anyone here runs `hx` frequently and/or for long-periods, I recommend `target-cpu=native`. I haven't tried it (yet), so I can't say if compile-times change considerably

**Reply by magistau** [reply]
> I don't think we're planning to do this since we want the plugin system to handle this kind of stuff.

I don't see how hx's questionable defaults are a job for third-party plugins.

**Comment by odurc** [comment]
Alternatively, to the original proposal, I'd suggest creating a configuration to set a default indent and unit **only for newly created buffers**, i.e. just after I hit `:n` + enter. The new buffer has no language associated with it yet and therefore it should simply use my default definition of indent and unit regardless of the language I'm writing. After I write the buffer (`:w`), hence naming the language via the file extension, Helix should continue to use whatever indent configuration it has started with regardless of the language which is the case already with "indent detection".

**Comment by odama626** [comment]
what about
```toml

[[language]]
name = "*"
````

at the top of your languages config will allow setting defaults to all languages options, and will be overridden by more specific names below it?

**Comment by Oglo12** [comment]
Something that would also be a good idea is to have some sort of default precedence since stuff like Makefiles require tabs and not spaces.

**Comment by literallylara** [comment]
Being able to set the tab-width globally is also an accessibility concern. Some people need a different tab width for better readability. This wouldn't change the code as it's just a visual setting. Having to change it for every language is very tedious.

Please reconsider.

**Comment by Oglo12** [comment]
And also the amount of spaces inserted. That is what I was getting at, but visual tab width too.

Sent from [Proton Mail](https://proton.me/mail/home) for Android.

-------- Original Message --------

**Comment by Kill-Anon** [comment]
Hi. I am new to Helix. I am deciding to migrate over from neovim, in search of creating a performative IDE for my new programming language. I might be under some misconceptions here, but from what I understand from reading this, talking with AI, and going through the documentation, that the developers prefer to let the language backends dictate the indentation settings for tab width and tabs vs spaces, requiring the end user to override in every particular situation. 

Re: **No global setting for preferring Tabs or Spaces**

- Reasoning seems sound enough.
- Whitespace sometimes dictates syntax.
- For example, I am writing a language that specifically requires tabs for indentation.

Ergo: 

- Language backend should be able to dictate tab vs spaces for indentation settings. 

However:

- It is annoying when not required for the syntax.
- Given that this is generally not the case that whitespace dictates syntax, there is far less effort involved on having users dictate the exceptions to a global rule, rather than prohibitting a global rule altogether.

Re: **No global setting for preferring Tab Width**

- Same as above.
- I was initially opposed, but have come to understand the reasoning behind the developers choices.
- I am writing a language that stylistically depends upon 8 spaced tab indentations, and it is nice for the backend I will develop to be able to override settings.
- I am just frustrated that I have spent several hours on trying to figure this out. ( And still havn't )

Suggestions:
-

1. Make the documentation clearer on Indentation. Dedicate an entire simplified section here https://docs.helix-editor.com/editor.html#editorindent-guides-section upon specifically indentation and tab width, and thats it. 

- In searching the documentation, I found things like https://docs.helix-editor.com/editor.html#editorindent-guides-section or https://docs.helix-editor.com/guides/indent.html are too verbose, unrelated, or esoteric to answer my question on how to setup global tab widths or indentations.

2. Unless indented whitespace serves as syntax, do not have the backends override global settings. There really should just be a boolean or flag that indicates whether a language has requisite demands for particular indentation settings.

**Reply by Kill-Anon** [reply]
I am extremely frustated. I have this written in my languages.toml file

```toml
[[language]]
name = "c"
scope = "source.c"
file-types = ["c", "h"]
indent = { tab-width = 8, unit = "\t" }
```

In spite of this, the file started off as indented with spaces, and when even after I set the `:indent-style` to "t", it displays as 4 spaces:

<img width="326" height="284" alt="image" src="https://github.com/user-attachments/assets/6e87497a-fcad-43fa-a140-584f39062c86" />

In my toml files though, it displays as 8 width like I asked. I am again spending a lot of time trying to figure out how to set this tab width. FWIW, AI has been of no help. It just tends to give me vim commands.

**Reply by sploders101** [reply]
Helix will also auto-detect what format is currently in use and follow that by default. Judging by the cursor in your screenshot, it looks like the file is using spaces instead of tabs, which is why it's not rendering at the correct width. Helix only modifies the width of tabs. Spaces are always one column.

**Reply by Kill-Anon** [reply]
Yeah I discovered later that a code generator, added a hidden configuration file that hx was deferring to.

**Comment by dancrossnyc** [comment]
This is one of my biggest frustrations with Helix, which I have more or less completely switched to otherwise.

It strikes me that an issue is that not all files conform to existing programming languages; some are simply free-form text. For example, I was editing `/etc/motd` on a shared Unix system earlier; that's approximately just a free-form text file. Or one might have a text file some free-form notes in it. Being able to specify tab widths for such things could be useful, yet apparently cannot even set the tab width from the command palette.

Some files containing configuration data aren't worth adding a dedicated language for; consider `/etc/ethers` (which still exists on some machines!): this is just a mapping from MAC addresses to host names; often people use tabs between fields just to keep things tidy, but it's so niche it's hardly worth the effort to do significant customization. Having some sort of global default that could be used for such things would be enormously useful.

Perhaps some sort of inheritance syntax could be leveraged to ameliorate some of the tedium of repeated language definitions, though I could imagine some thorny issues when it comes to overriding settings that a different language inherits from; to my knowledge there is nothing like that presently, though `.editorconfig` seems to come (kind of) close.
